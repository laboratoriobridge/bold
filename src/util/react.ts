import { MutableRefObject, Ref } from 'react'

export function setRef<T>(ref: Ref<T>, value: any) {
    if (typeof ref === 'function') {
        ref(value)
    } else if (ref) {
        (ref as MutableRefObject<T>).current = value
    }
}
