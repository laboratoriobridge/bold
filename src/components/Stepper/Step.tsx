import React, { CSSProperties } from 'react'

import { isNil } from 'lodash'
import { ExternalStyles, focusBoxShadow, Theme, useStyles } from '../../styles'
import { Omit } from '../../util'
import { getComponents } from '../../util/overrides'
import CheckIcon from '../Icon/generated/CheckDefault'
import MinusCircleFilled from '../Icon/generated/MinusCircleFilled'

export type StepStatus = 'active' | 'completed' | 'incompleted' | 'inactive'

export interface StepProps extends Omit<React.HTMLAttributes<HTMLSpanElement>, 'style'> {
  status?: StepStatus
  hasConnector?: boolean
  style?: ExternalStyles
  overrides?: {
    Root?: React.ElementType
    Connector?: React.ElementType
    IconContainer?: React.ElementType
    Icon?: React.ElementType
    Label?: React.ElementType
  }
}

export function Step(props: StepProps) {
  const { status, hasConnector, style, overrides, children, ...rest } = props
  const { Root, Icon, IconContainer, Connector, Label } = getComponents(overrides, defaultComponents)
  const { classes, css } = useStyles(createStyles, props)

  const hasBackground =
    status === 'active' || status === 'completed' || (status === 'incompleted' && isNil(overrides?.Icon))

  return (
    <Root className={css(classes.step, style)} {...rest}>
      {hasConnector && <Connector className={css(classes.connector)} />}

      <IconContainer className={classes.container}>
        {hasBackground && <div className={classes.background} />}
        {Icon && <Icon className={classes.icon} />}
        {!Icon && status === 'completed' && <CheckIcon className={classes.icon} />}
        {!Icon && status === 'inactive' && <MinusCircleFilled className={classes.icon} />}
      </IconContainer>

      <Label className={classes.label}>{children}</Label>
    </Root>
  )
}

Step.defaultProps = {
  status: 'incompleted',
  hasConnector: true,
} as Partial<StepProps>

export const defaultComponents: StepProps['overrides'] = {
  Root: 'span',
  Icon: null,
  IconContainer: 'span',
  Connector: 'span',
  Label: 'span',
}

const createStyles = (theme: Theme, { status }: StepProps) => {
  return {
    step: {
      position: 'relative',
      flex: '1',
      display: 'flex',
      flexDirection: 'column',
      alignItems: 'center',
      textAlign: 'center',
      padding: '0 0.5rem',
      gap: '0.75rem',
    } as CSSProperties,
    connector: {
      position: 'absolute',
      top: 'calc(0.75rem - 1px)',
      left: 'calc(-50% + 0.5rem)',
      right: 'calc(50% + 0.5rem)',
      borderTopWidth: '2px',
      borderTopStyle: status === 'inactive' ? 'dashed' : 'solid',
      borderTopColor:
        status === 'incompleted' || status === 'inactive' ? theme.pallete.gray.c80 : theme.pallete.primary.main,
      transition: 'all .4s ease',
    } as CSSProperties,
    container: {
      zIndex: 1,
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'center',
      width: '1.5rem',
      height: '1.5rem',
      textAlign: 'center',
      position: 'relative',
    } as CSSProperties,
    background: {
      width: '1rem',
      height: '1rem',
      borderRadius: '50%',
      position: 'absolute',
      top: 'calc(50% - 0.5rem)',
      left: 'calc(50% - 0.5rem)',
      background: status === 'incompleted' ? theme.pallete.gray.c60 : theme.pallete.primary.main,
      boxShadow:
        (status === 'active' && focusBoxShadow(theme, 'primary')) ||
        (status === 'completed' && `0 0 0 4px ${theme.pallete.primary.main}`),
      transition: 'all .4s ease',
    } as CSSProperties,
    icon: {
      zIndex: 1,
      fill: status === 'completed' || status === 'active' ? theme.pallete.primary.c100 : theme.pallete.gray.c60,
      width: '1.25rem',
      height: '1.25rem',
    } as CSSProperties,
    label: {
      color: getLabelColor(status, theme),
      fontWeight: 'bold',
      transition: 'all .4s ease',
    } as CSSProperties,
  }
}

const getLabelColor = (status: StepStatus, theme: Theme): CSSProperties['color'] => {
  if (status === 'active') return theme.pallete.primary.main
  if (status === 'inactive') return theme.pallete.text.disabled
  return theme.pallete.text.main
}
