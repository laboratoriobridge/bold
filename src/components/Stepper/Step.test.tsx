import { render } from '@testing-library/react'
import React from 'react'

import { Step } from './Step'

it('should render correctly when status is incompleted', () => {
  const { container } = render(<Step status='incompleted'>Incompleted step</Step>)
  expect(container).toMatchSnapshot()
})

it('should render correctly when status is completed', () => {
  const { container } = render(<Step status='completed'>Completed step</Step>)
  expect(container).toMatchSnapshot()
})

it('should render correctly when status is active', () => {
  const { container } = render(<Step status='active'>Active step</Step>)
  expect(container).toMatchSnapshot()
})

it('should render correctly when status is inactive', () => {
  const { container } = render(<Step status='inactive'>Inactive step</Step>)
  expect(container).toMatchSnapshot()
})

it('should accept "style" prop', () => {
  const { container } = render(<Step style={{ color: 'red' }}>Step red</Step>)
  expect(container).toMatchSnapshot()
})

it('should allow override of components', () => {
  const RootOverride = (props) => <span id='root' {...props} />
  const ConnectorOverride = (props) => <span id='connector' {...props} />
  const IconOverride = (props) => <span id='icon' {...props} />
  const IconContainerOverride = (props) => <span id='icon-container' {...props} />
  const LabelOverride = (props) => <span id='label' {...props} />

  const { container } = render(
    <Step
      overrides={{
        Root: RootOverride,
        Connector: ConnectorOverride,
        Icon: IconOverride,
        IconContainer: IconContainerOverride,
        Label: LabelOverride,
      }}
    >
      Step with overrides
    </Step>
  )
  expect(container).toMatchSnapshot()
})

it('should not render Connector when "hasConnector" props is false', () => {
  const Connector = (props) => <span id='connector' />

  const { rerender, container } = render(
    <Step overrides={{ Connector }} hasConnector={false}>
      Step with no connector
    </Step>
  )
  expect(container.querySelector('#connector')).toBeFalsy()

  rerender(
    <Step overrides={{ Connector }} hasConnector>
      Step with no connector
    </Step>
  )
  expect(container.querySelector('#connector')).toBeTruthy()
})
