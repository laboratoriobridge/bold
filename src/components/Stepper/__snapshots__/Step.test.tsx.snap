// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should accept "style" prop 1`] = `
.emotion-0 {
  position: absolute;
  top: calc(0.75rem - 1px);
  left: calc(-50% + 0.5rem);
  right: calc(50% + 0.5rem);
  border-top-width: 2px;
  border-top-style: solid;
  border-top-color: #D3D4DD;
  -webkit-transition: all .4s ease;
  transition: all .4s ease;
}

.emotion-2 {
  z-index: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 1.5rem;
  height: 1.5rem;
  text-align: center;
  position: relative;
}

.emotion-1 {
  width: 1rem;
  height: 1rem;
  border-radius: 50%;
  position: absolute;
  top: calc(50% - 0.5rem);
  left: calc(50% - 0.5rem);
  background: #8F8FA2;
  -webkit-transition: all .4s ease;
  transition: all .4s ease;
}

.emotion-3 {
  color: #24252E;
  font-weight: bold;
  -webkit-transition: all .4s ease;
  transition: all .4s ease;
}

.emotion-4 {
  position: relative;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  padding: 0 0.5rem;
  gap: 0.75rem;
  color: red;
}

<div>
  <span
    class="emotion-4"
  >
    <span
      class="emotion-0"
    />
    <span
      class="emotion-2"
    >
      <div
        class="emotion-1"
      />
    </span>
    <span
      class="emotion-3"
    >
      Step red
    </span>
  </span>
</div>
`;

exports[`should allow override of components 1`] = `
.emotion-4 {
  position: relative;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  padding: 0 0.5rem;
  gap: 0.75rem;
}

.emotion-0 {
  position: absolute;
  top: calc(0.75rem - 1px);
  left: calc(-50% + 0.5rem);
  right: calc(50% + 0.5rem);
  border-top-width: 2px;
  border-top-style: solid;
  border-top-color: #D3D4DD;
  -webkit-transition: all .4s ease;
  transition: all .4s ease;
}

.emotion-2 {
  z-index: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 1.5rem;
  height: 1.5rem;
  text-align: center;
  position: relative;
}

.emotion-1 {
  z-index: 1;
  fill: #8F8FA2;
  width: 1.25rem;
  height: 1.25rem;
}

.emotion-3 {
  color: #24252E;
  font-weight: bold;
  -webkit-transition: all .4s ease;
  transition: all .4s ease;
}

<div>
  <span
    class="emotion-4"
    id="root"
  >
    <span
      class="emotion-0"
      id="connector"
    />
    <span
      class="emotion-2"
      id="icon-container"
    >
      <span
        class="emotion-1"
        id="icon"
      />
    </span>
    <span
      class="emotion-3"
      id="label"
    >
      Step with overrides
    </span>
  </span>
</div>
`;

exports[`should render correctly when status is active 1`] = `
.emotion-4 {
  position: relative;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  padding: 0 0.5rem;
  gap: 0.75rem;
}

.emotion-2 {
  z-index: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 1.5rem;
  height: 1.5rem;
  text-align: center;
  position: relative;
}

.emotion-0 {
  position: absolute;
  top: calc(0.75rem - 1px);
  left: calc(-50% + 0.5rem);
  right: calc(50% + 0.5rem);
  border-top-width: 2px;
  border-top-style: solid;
  border-top-color: #0069D0;
  -webkit-transition: all .4s ease;
  transition: all .4s ease;
}

.emotion-1 {
  width: 1rem;
  height: 1rem;
  border-radius: 50%;
  position: absolute;
  top: calc(50% - 0.5rem);
  left: calc(50% - 0.5rem);
  background: #0069D0;
  box-shadow: 0 0 0 2px #F0F0F5,0 0 0 4px #0069D0;
  -webkit-transition: all .4s ease;
  transition: all .4s ease;
}

.emotion-3 {
  color: #0069D0;
  font-weight: bold;
  -webkit-transition: all .4s ease;
  transition: all .4s ease;
}

<div>
  <span
    class="emotion-4"
  >
    <span
      class="emotion-0"
    />
    <span
      class="emotion-2"
    >
      <div
        class="emotion-1"
      />
    </span>
    <span
      class="emotion-3"
    >
      Active step
    </span>
  </span>
</div>
`;

exports[`should render correctly when status is completed 1`] = `
.emotion-5 {
  position: relative;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  padding: 0 0.5rem;
  gap: 0.75rem;
}

.emotion-3 {
  z-index: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 1.5rem;
  height: 1.5rem;
  text-align: center;
  position: relative;
}

.emotion-4 {
  color: #24252E;
  font-weight: bold;
  -webkit-transition: all .4s ease;
  transition: all .4s ease;
}

.emotion-0 {
  position: absolute;
  top: calc(0.75rem - 1px);
  left: calc(-50% + 0.5rem);
  right: calc(50% + 0.5rem);
  border-top-width: 2px;
  border-top-style: solid;
  border-top-color: #0069D0;
  -webkit-transition: all .4s ease;
  transition: all .4s ease;
}

.emotion-1 {
  width: 1rem;
  height: 1rem;
  border-radius: 50%;
  position: absolute;
  top: calc(50% - 0.5rem);
  left: calc(50% - 0.5rem);
  background: #0069D0;
  box-shadow: 0 0 0 4px #0069D0;
  -webkit-transition: all .4s ease;
  transition: all .4s ease;
}

.emotion-2 {
  z-index: 1;
  fill: #FFFFFF;
  width: 1.25rem;
  height: 1.25rem;
}

<div>
  <span
    class="emotion-5"
  >
    <span
      class="emotion-0"
    />
    <span
      class="emotion-3"
    >
      <div
        class="emotion-1"
      />
      <svg
        class="emotion-2"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          clip-rule="evenodd"
          d="M18.294 6.293a1 1 0 111.414 1.414l-9 9a1 1 0 01-1.414 0l-5.001-5.001a1 1 0 011.414-1.414l4.294 4.294 8.293-8.293z"
          fill-rule="evenodd"
        />
      </svg>
    </span>
    <span
      class="emotion-4"
    >
      Completed step
    </span>
  </span>
</div>
`;

exports[`should render correctly when status is inactive 1`] = `
.emotion-4 {
  position: relative;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  padding: 0 0.5rem;
  gap: 0.75rem;
}

.emotion-2 {
  z-index: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 1.5rem;
  height: 1.5rem;
  text-align: center;
  position: relative;
}

.emotion-1 {
  z-index: 1;
  fill: #8F8FA2;
  width: 1.25rem;
  height: 1.25rem;
}

.emotion-0 {
  position: absolute;
  top: calc(0.75rem - 1px);
  left: calc(-50% + 0.5rem);
  right: calc(50% + 0.5rem);
  border-top-width: 2px;
  border-top-style: dashed;
  border-top-color: #D3D4DD;
  -webkit-transition: all .4s ease;
  transition: all .4s ease;
}

.emotion-3 {
  color: #AAAAB9;
  font-weight: bold;
  -webkit-transition: all .4s ease;
  transition: all .4s ease;
}

<div>
  <span
    class="emotion-4"
  >
    <span
      class="emotion-0"
    />
    <span
      class="emotion-2"
    >
      <svg
        class="emotion-1"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M2 12C2 6.477 6.477 2 12 2s10 4.477 10 10c0 5.522-4.477 10-10 10S2 17.522 2 12zm14-1H8v2h8v-2z"
        />
      </svg>
    </span>
    <span
      class="emotion-3"
    >
      Inactive step
    </span>
  </span>
</div>
`;

exports[`should render correctly when status is incompleted 1`] = `
.emotion-4 {
  position: relative;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  padding: 0 0.5rem;
  gap: 0.75rem;
}

.emotion-0 {
  position: absolute;
  top: calc(0.75rem - 1px);
  left: calc(-50% + 0.5rem);
  right: calc(50% + 0.5rem);
  border-top-width: 2px;
  border-top-style: solid;
  border-top-color: #D3D4DD;
  -webkit-transition: all .4s ease;
  transition: all .4s ease;
}

.emotion-2 {
  z-index: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 1.5rem;
  height: 1.5rem;
  text-align: center;
  position: relative;
}

.emotion-1 {
  width: 1rem;
  height: 1rem;
  border-radius: 50%;
  position: absolute;
  top: calc(50% - 0.5rem);
  left: calc(50% - 0.5rem);
  background: #8F8FA2;
  -webkit-transition: all .4s ease;
  transition: all .4s ease;
}

.emotion-3 {
  color: #24252E;
  font-weight: bold;
  -webkit-transition: all .4s ease;
  transition: all .4s ease;
}

<div>
  <span
    class="emotion-4"
  >
    <span
      class="emotion-0"
    />
    <span
      class="emotion-2"
    >
      <div
        class="emotion-1"
      />
    </span>
    <span
      class="emotion-3"
    >
      Incompleted step
    </span>
  </span>
</div>
`;
